File: src/sentry.zig
const mode: std.builtin.Mode = @import("builtin").mode;
const is_test: bool = @import("builtin").is_test;
pub const sentry_on_crashed_last_run_callback = ?fn (?*anyopaque) callconv(.C) void;
pub const sentry_before_send_event_callback = ?fn (?*anyopaque) callconv(.C) void;
pub const struct_sentry_options_t = extern struct {
    enabled: bool = true,
    debug: bool = mode == .debug or is_test,
    enable_out_of_memory_tracking: bool = true,
    enable_swizzling: bool = false,
    dsn: [*:0]const u8 = "",
    release_name: [*:0]const u8 = "",
    on_before_send: sentry_before_send_event_callback = null,
    on_crashed_last_run: sentry_on_crashed_last_run_callback = null,
};
pub const sentry_options_t = struct_sentry_options_t;
pub extern fn sentry_startSentry(options: *const sentry_options_t) void;
pub extern fn sentry_captureMessage(msg: [*:0]const u8) void;
pub extern fn sentry_stopSentry() void;
pub extern fn sentry_didCrashLastRun() bool;
pub extern fn sentry_isEnabled() bool;
pub extern fn sentry_crash() void;

const std = @import("std");

pub const Sentry = struct {
    pub inline fn start(options: struct_sentry_options_t) void {
        var opts = options;
        sentry_startSentry(&opts);
    }

    pub inline fn captureMessage(msg: [:0]const u8) void {
        std.debug.assert(sentry_isEnabled());
        sentry_captureMessage(msg);
    }

    pub inline fn stop() void {
        sentry_stopSentry();
    }

    pub inline fn crash() void {
        sentry_crash();
    }

    pub inline fn didCrashLastRun() bool {
        return sentry_didCrashLastRun();
    }

    pub inline fn isEnabled() bool {
        return sentry_isEnabled();
    }
};

test "Sentry start" {
    Sentry.start(struct_sentry_options_t{
        .enabled = true,
        .debug = true,
        .release_name = "sentry-cocoa-zig-test-1.0",
        // .dsn = "https://3bfc34ee0b624366be3b918bfcdf25cd:24fd460f12b84e01838032ee5bf9a0ee@o244632.ingest.sentry.io/6133230",
    });
}

test "Sentry crash" {
    Sentry.start(struct_sentry_options_t{
        .enabled = true,
        .debug = false,
        .release_name = "sentry-cocoa-zig-test-1.0",
        // .dsn = "https://3bfc34ee0b624366be3b918bfcdf25cd:24fd460f12b84e01838032ee5bf9a0ee@o244632.ingest.sentry.io/6133230",
    });
    Sentry.captureMessage("Hello");
}

File: src/sentry.m

#include "sentry.h"
#include <Foundation/Foundation.h>
#include <sentry/Sentry.h>

static sentry_before_send_event_callback _on_before_send_callback;
static sentry_on_crashed_last_run_callback _on_crashed_last_run_callback;
void sentry_startSentry(const sentry_options_t *options) {
  SentryOptions *_options = [[SentryOptions alloc] init];
  [_options setEnabled:options->enabled];
  [_options setDebug:options->debug];
  _options.dsn = [NSString stringWithUTF8String:options->dsn];
  [_options setEnableSwizzling:false];
  [_options setEnableAutoSessionTracking:false];
  [_options setEnableAutoPerformanceTracking:false];
  [_options setEnableNetworkTracking:false];
  [_options setEnableOutOfMemoryTracking:false];

  [_options
      setReleaseName:[NSString stringWithUTF8String:options->release_name]];

  if (options->on_before_send) {
    // copy it so we don't worry about allocation
    _on_before_send_callback = options->on_before_send;
    _options.beforeSend = ^(SentryEvent *event) {
      if (_on_before_send_callback) {
        _on_before_send_callback(event);
      }

      return event;
    };
  }

  if (options->on_crashed_last_run) {
    // copy it so we don't worry about allocation
    _on_crashed_last_run_callback = options->on_crashed_last_run;
    _options.onCrashedLastRun = ^(SentryEvent *event) {
      if (_on_crashed_last_run_callback) {
        _on_crashed_last_run_callback();
      }
    };
  }

  [SentrySDK startWithOptionsObject:_options];
  [SentrySDK startSession];
}
void sentry_stopSentry() { [SentrySDK close]; }
bool sentry_didCrashLastRun() { return [SentrySDK crashedLastRun]; }
void sentry_crash() { [SentrySDK crash]; }
bool sentry_isEnabled() { return [SentrySDK isEnabled]; }

void sentry_captureMessage(const char *msg) {
  [SentrySDK captureMessage:[NSString stringWithUTF8String:msg]];
}